From 0fb964a38c43874001c61d99ff44ab08a53bb5ea Mon Sep 17 00:00:00 2001
From: Xinyu Chen <xinyu.chen@freescale.com>
Date: Fri, 15 Aug 2008 13:18:25 +0800
Subject: [PATCH] ENGR00085834 MX31 TVout failed after resuming from power suspend.

Add TV Encoder power status check in suspend
And do not enable TV Encoder in resume if disable before suspend

Signed-off-by: Xinyu Chen <xinyu.chen@freescale.com>
(cherry picked from commit 920df75bd09dbed8084d3767e62ced53cf81c9f7)
---
 drivers/video/mxc/ch7024.c |   16 +++++++++++-----
 1 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/video/mxc/ch7024.c b/drivers/video/mxc/ch7024.c
index eff448f..d8e3d90 100644
--- a/drivers/video/mxc/ch7024.c
+++ b/drivers/video/mxc/ch7024.c
@@ -146,6 +146,7 @@ struct ch7024_input_timing {
 #define TVOUT_FMT_PAL	2
 
 static int enabled;		/* enable power on or not */
+static int pm_status;		/* status before suspend */
 
 static struct i2c_client *ch7024_client;
 static struct fb_info *ch7024_fbi;
@@ -805,17 +806,22 @@ static int ch7024_remove(struct i2c_client *client)
 static int ch7024_suspend(struct i2c_client *client, pm_message_t state)
 {
 	pr_debug("Ch7024 suspend routing..\n");
-	ch7024_disable();
+	if (enabled) {
+		ch7024_disable();
+		pm_status = 1;
+	} else {
+		pm_status = 0;
+	}
 	return 0;
 }
 
 static int ch7024_resume(struct i2c_client *client)
 {
 	pr_debug("Ch7024 resume routing..\n");
-	ch7024_enable();
-	ch7024_write_reg(CH7024_RESET, 0x0);
-	ch7024_write_reg(CH7024_RESET, 0x3);
-	ch7024_write_reg(CH7024_POWER, 0x0D);
+	if (pm_status) {
+		ch7024_enable();
+		ch7024_setup(ch7024_cur_mode);
+	}
 	return 0;
 }
 #else
-- 
1.5.4.4

